### PR TITLE
[Wait4PR#671][ Layer ] Add Concat Layer @open sesame 10/28 14:16

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR includes Concatenate layer.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/src/concat_layer.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    concat_channel += dim.channel();
+  }
+
+  if (concat_channel != output_dim[0].channel()) {
+    throw std::runtime_error(
+      "Error: Sum of channel of input layers is not same with output channel");
+  }
+
+  hidden = Tensor(output_dim[0]);
+  hidden.setZero();
```
Comment (kparichay) [inline nntrainer/src/concat_layer.cpp:64]: this is redundant

File: nntrainer/src/concat_layer.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+sharedConstTensors ConcatLayer::forwarding(sharedConstTensors in) {
+
+  TensorDim d = in[0]->getDim();
+  unsigned int concat_channel = d.channel();
+
+  for (unsigned int idx = 1; idx < num_inputs; ++idx) {
+    TensorDim dim = in[idx]->getDim();
+
+    for (unsigned int i = 2; i < d.rank(); ++i) {
+      if (d[i] != dim[i])
```
Comment (kparichay) [inline nntrainer/src/concat_layer.cpp:51]: how about moving these rank dependent checks to initialise with input_dim and rather just compare in[i].getDim() with input_dim[i]? lets keep forwarding (critical path) least checks.

Comment (kparichay) [reply nntrainer/src/concat_layer.cpp:51]: Further, I think that these checks of the dimension of `in` and `input_dim` can be all skipped from `forwarding` as nntrainer layers passes the data to this layer. If input to the model is correct dimension, this function will automatically receive the correct dimensions. This is the reason no other layer checks the dimension of the input in forwarding/backwarding.

Comment (zhoonit) [reply nntrainer/src/concat_layer.cpp:51]: Maybe it could live inside `#ifdef DEBUG`

File: nntrainer/src/concat_layer.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+  return ret;
+}
+
+void ConcatLayer::setProperty(const PropertyType type,
+                              const std::string &value) {
+  int status = ML_ERROR_NONE;
+
+  switch (type) {
+  case PropertyType::num_inputs: {
```
Comment (zhoonit) [inline nntrainer/src/concat_layer.cpp:110]: IMO, it is more natural to set num_inputs from parsing input_layers in and disallow setting for `num_inputs`

```ini
[concat1]
Type = concat
input_layers ="input_layer1, input_layer2"
```

Comment (jijoongmoon) [reply nntrainer/src/concat_layer.cpp:110]: This num_inputs may not required according to following implementation. I will change this as it implemented.

File: nntrainer/src/concat_layer.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/src/concat_layer.cpp:52]: fixed. Thanks

File: nntrainer/src/concat_layer.cpp (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/src/concat_layer.cpp]: fixed. Thanks

File: nntrainer/src/concat_layer.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    for (unsigned int i = 2; i < d.rank(); ++i) {
+      if (d[i] != dim[i])
+        throw std::runtime_error("Error: concat layer requires same "
+                                 "shape from  all input layers");
+    }
+#endif
+    concat_channel += dim.channel();
+  }
+
+  if (concat_channel != output_dim[0].channel()) {
```
Comment (kparichay) [inline nntrainer/src/concat_layer.cpp:60]: this check and computation should also move to DEBUG.

File: nntrainer/src/concat_layer.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+sharedConstTensors ConcatLayer::backwarding(sharedConstTensors derivative,
+                                            int iteration) {
+  sharedConstTensors ret;
+  TensorDim d = derivative[0]->getDim();
+
+  unsigned int position = 0;
+  for (unsigned int idx = 0; idx < num_inputs; ++idx) {
+    TensorDim in_dim = input_dim[idx];
+    sharedTensor t = std::shared_ptr<Tensor>(new Tensor(in_dim),
+                                             std::default_delete<Tensor[]>());
```
Comment (kparichay) [inline nntrainer/src/concat_layer.cpp:92]: I am not if this should call `delete Tensor[]` than `delete Tensor`.

File: nntrainer/src/concat_layer.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/src/concat_layer.cpp:52]: @jijoongmoon I think this check is still needed at the initialize phase to ensure that the connected layers set by the model are compatible.

File: nntrainer/src/concat_layer.cpp (hunk 5 above)
Comment (jijoongmoon) [reply nntrainer/src/concat_layer.cpp]: fixed. Thanks

File: nntrainer/src/concat_layer.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/src/concat_layer.cpp:61]: fixed. Thanks

**Review by kparichay — APPROVED** [review]
LGTM

**Review by zhoonit — APPROVED** [review]
LGTM :)

**Pull request merged**